Selection algorithms (#500)
fortran-lang/stdlib: doc/specs/stdlib_selection.md

were derived using some code from quickSelect in the Coretran library, by Leon Foks,
https://github.com/leonfoks/coretran. Leon Foks has given permission for the code here
to be released under stdlib's MIT license.

## `select` - find the k-th smallest value in an input array

### Status

Experimental

### Description

Returns the k-th smallest value of `array(:)`, and also partially sorts `array(:)`
such that `all(array(1:k) <= array(k))`  and `all(array(k) <= array((k+1):size(array)))`

### Syntax

`call [[stdlib_selection(module):select(interface)]]( array, k, kth_smallest [, left, right ] )`

### Class

Generic subroutine.

### Arguments

`array` : shall be a rank one array of any of the types:
`integer(int8)`, `integer(int16)`, `integer(int32)`, `integer(int64)`,
`real(sp)`, `real(dp)`, `real(qp)`. It is an `intent(inout)` argument. 

`k`: shall be a scalar with any of the types:
`integer(int8)`, `integer(int16)`, `integer(int32)`, `integer(int64)`. It
is an `intent(in)` argument. We search for the `k`-th smallest entry of `array(:)`.

`kth_smallest`: shall be a scalar with the same type as `array`. On return it contains

> the k-th smallest entry of `array(:)`.

Reviewer: The type of `intent` is missing

Author: Good catch -- fixed in a commit just now.

---
title: Selection Procedures
---

# The `stdlib_selection` module

[TOC]

## Overview of selection

Suppose you wish to find the value of the k-th smallest entry in an array of size N, or
the index of that value. While it could be done by sorting the whole array
using `[[stdlib_sorting(module):sort(interface)]]` or 
`[[stdlib_sorting(module):sort_index(interface)]]` from 
`[[stdlib_sorting(module)]]` and then finding the k-th entry, that would
require O(N x LOG(N)) time. However selection of a single entry can be done in
O(N) time, which is much faster for large arrays.  This is useful, for example,
to quickly find the median of an array, or some other percentile.

The Fortran Standard Library therefore provides a module, `stdlib_selection`,
which implements selection algorithms.

## Overview of the module

The module `stdlib_selection` defines two generic subroutines:
* `select` is used to find the k-th smallest entry of an array. The input
array is also modified in-place, and on return will be partially sorted

> such that `all(array(1:k) <= array(k)))`  and `all(array(k) <= array((k+1):size(array)))`.

Reviewer 2: ```suggestion
such that `all(array(1:k) <= array(k)))`  and `all(array(k) <= array((k+1):size(array)))` is true.
```

for the k-th smallest value. This can be useful if you have previously called `select`
to find a smaller or larger rank (that will have led to partial sorting of
`array`, thus implying some constraints on the location).

* `arg_select` is used to find the index of the k-th smallest entry of an array.
In this case the input array is not modified, but the user must provide an
input index array with the same size as `array`, having indices that are a permutation of
`1:size(array)`, which is modified instead. On return the index array is modified
such that `all(array(index(1:k)) <= array(index(k)))` and `all(array(k) <= array(k+1:size(array)))`.
The user can optionally specify `left` and `right` indices to constrain the search
for the k-th smallest value. This can be useful if you have previously called `arg_select`
to find a smaller or larger rank (that will have led to partial sorting of
`index`, thus implying some constraints on the location).

## `select` - find the k-th smallest value in an input array

### Status

Experimental

### Description

Returns the k-th smallest value of `array(:)`, and also partially sorts `array(:)`

> such that `all(array(1:k) <= array(k))`  and `all(array(k) <= array((k+1):size(array)))`

Reviewer 2: A line mentioning if the sort is stable or not would be helpful.

Author: To my understanding the concept of stable sort does not apply here? Since we are sorting based on the value of `array` itself, not via a comparison function or some other indirect method. 

In other words, if two values of `array` are equal, they really are indistinguishable (in all contexts).

Author: But the point does hold for `arg_select` -- and I am not immediately sure whether it is stable, will have to look into it.

Author: From experiments, `arg_select` does not produce a stable sort. I have mentioned that in the documentation.

such that `all(array(1:k) <= array(k)))`  and `all(array(k) <= array((k+1):size(array)))`.
The user can optionally specify `left` and `right` indices to constrain the search
for the kth-smallest value. This can be useful if you have previously called `select`
to find a smaller or larger rank (that will have led to partial sorting of
`array`, thus implying some constraints on the location).

* `arg_select` is used to find the index of the kth-smallest entry of an array.
In this case the input array is not modified, but the user must provide an
input index array with the same size as `array`, having unique indices from
`1:size(array)`, which is modified instead. On return the index array is modified
such that `all(array(index(1:k)) <= array(index(k)))` and `all(array(k) <= array(k+1:size(array)))`.
The user can optionally specify `left` and `right` indices to constrain the search
for the kth-smallest value. This can be useful if you have previously called `arg_select`
to find a smaller or larger rank (that will have led to partial sorting of
`index`, thus implying some constraints on the location).

#### Licensing

The Fortran Standard Library is distributed under the MIT

> License. However components of the library may be based on code with
> additional licensing restrictions. In particular `select` and `arg_select`

Reviewer: Based on the following sentences, there are no additional restrictions compared to MIT license.
Therefore I propose the following sentences:

```suggestion
License. It is worth noting that  `select` and `arg_select`
```

https://github.com/leonfoks/coretran. Leon Foks has given permission for the code here

> to be released under stdlib's MIT license.

Reviewer 3: Is this subsection necessary? If the permission was granted to port the code to stdlib's MIT license, then the text in the LICENSE file conveys all the information.

Author: Agree it is no longer necessary -- so I removed this section in a recent commit. 

However I wanted to keep acknowledgement of Leon Fok's contribution, and so I have edited the "Notes" sections to do that in the commit just now.

> https://github.com/leonfoks/coretran. Leon has given permission for the code here

Reviewer: ```suggestion
https://github.com/leonfoks/coretran. Leon has given permission for the code used in `stdlib`
```

> to be released under stdlib's MIT licence.

Reviewer: ```suggestion
to be released under stdlib's MIT license.
```

Reviewer: I am not a jurist. I guess it is ok as is.

### Specifications of the `stdlib_selection` procedures

> #### `select` - find the kth smallest value in an input array

Reviewer: ```suggestion
#### `select` - find the k-th smallest value in an input array
```

##### Status

Experimental

##### Description

Returns the k-th smallest value of `array(:)`, and also partially sorts `array(:)`
such that `all(array(1:k) <= array(k))`  and `all(array(k) <= array((k+1):size(array)))`

##### Syntax

`call [[stdlib_selection(module):select(interface)]]( array, k, kth_smallest [, left, right ] )`

##### Class

Generic subroutine.

##### Arguments

`array` : shall be a rank one array of any of the types:
`integer(int8)`, `integer(int16)`, `integer(int32)`, `integer(int64)`,
`real(sp)`, `real(dp)`, `real(qp)`. It is an `intent(inout)` argument. 

`k`: shall be a scalar with any of the types:
`integer(int8)`, `integer(int16)`, `integer(int32)`, `integer(int64)`. It
is an `intent(in)` argument. We search for the `k`-th smallest entry of `array(:)`.

`kth_smallest`: shall be a scalar with the same type as `array`. On return it contains
the k-th smallest entry of `array(:)`.

`left` (optional): shall be a scalar with the same type as `k`. It is an `intent(in)`
argument. If specified then we assume the k-th smallest value is definitely contained

> in `array(left:size(array))`. If not present it is 1. This is typically useful if multiple calls

Reviewer: ```suggestion
in `array(left:size(array))`. If `left` is not present, the default is 1. This is typically useful if multiple calls
```

to `select` are made, because the partial sorting of `array` implies constraints on where
we need to search.

`right` (optional): shall be a scalar with the same type as `k`. It is an `intent(in)`
argument. If specified then we assume the k-th smallest value is definitely contained

> in `array(1:right)`. If not present it is `size(array)`. This is typically useful if multiple calls

Reviewer: ```suggestion
in `array(1:right)`. If `right` is not present, the default is `size(array)`. This is typically useful if multiple calls
```

to `select` are made, because the partial sorting of `array` implies constraints on where
we need to search.

##### Notes

Selection of a single value should have runtime of O(`size(array)`), so it is

> asymptotically faster than sorting `array` entirely. The test program at the the

Reviewer: ```suggestion
asymptotically faster than sorting `array` entirely. The test program at the
```

end of this document shows that is the case.

### Example

```fortran

>     program demo_select

Reviewer 3: Here, and in other examples: Why the 4-space indent for the whole example? It seems unnecessary to me, and it will wrap the text on small screens (smartphones).

Author: Good point -- there is no reason -- so I removed the outer 4-space-indent in the commit just now.

> `all( array(1:k-1) <= array(k) )` and `all(array(k) <= array(k+1:size(array)))`.

Reviewer: Already mentioned in the description. Therefore I think it could be removed.

Author: agree, implemented in upcoming commits.

`array` : shall be a rank one array of any of the types:
`integer(int8)`, `integer(int16)`, `integer(int32)`, `integer(int64)`,
`real(sp)`, `real(dp)`, `real(qp)`. It is an `intent(in)` argument. On input it is
the array in which we search for the kth smallest entry.

`indx`: shall be a rank one array with the same size as `array`, containing all integers
from `1:size(array)` in any order. It is of any of the types:
`integer(int8)`, `integer(int16)`, `integer(int32)`, `integer(int64)`. It is an
`intent(inout)` argument. On return its elements will define a partial sorting of `array(:)` such that:
 `all( array(indx(1:k-1)) <= array(indx(k)) )` and `all(array(indx(k)) <= array(indx(k+1:size(array))))`.

`k`: shall be a scalar with the same type as `indx`. It is an `intent(in)`
argument. We search for the `k`-th smallest entry of `array(:)`.

`kth_smallest`: a scalar with the same type as `indx`. It is an `intent(out)` argument,
and on return it contains the index of the k-th smallest entry of `array(:)`.

`left` (optional): shall be a scalar with the same type as `k`. It is an `intent(in)`
argument. If specified then we assume the k-th smallest value is definitely contained

> in `array(indx(left:size(array)))`. If not present it is 1. This is typically useful if multiple calls

Reviewer: ```suggestion
in `array(indx(left:size(array)))`. If `left` is not present, the default is 1. This is typically useful if multiple calls
```

to `arg_select` are made, because the partial sorting of `indx` implies constraints on where
we need to search.

`right` (optional): shall be a scalar with the same type as `k`. It is an `intent(in)`
argument. If specified then we assume the k-th smallest value is definitely contained

> in `array(indx(1:right))`. If not present it is `size(array)`. This is typically useful if multiple calls

Reviewer: ```suggestion
in `array(indx(1:right))`. If `right` is not present, the default is `size(array)`. This is typically useful if multiple calls
```

to `arg_select` are made, because the reordering of `indx` implies constraints on
where we need to search.

##### Notes

`arg_select` does not modify `array`, unlike `select`.

While it is essential that that `indx` contains the integers `1:size(array)` (in any

> order), the code does not check for this.

Reviewer: I am not sure to understand what the code doesn't check.

Author: clarified in an upcoming commit.
```
While it is essential that that `indx` contains a permutation of the integers `1:size(array)`, 
the code does not check for this. For example if `size(array) == 4`, then we could have 
`indx = [4, 2, 1, 3]` or `indx = [1, 2, 3, 4]`, but not `indx = [2, 1, 2, 4]`. It is the user's 
responsibility to avoid such errors.
```

asymptotically faster than sorting `array` entirely. The test program at the end of
these documents confirms that is the case.

##### Example

```fortran
    program demo_arg_select
      use stdlib_selection, only: arg_select
      implicit none

      real, allocatable :: array(:)
      integer, allocatable :: indx(:)
      integer :: kth_smallest
      integer :: k, left, right

      array = [3., 2., 7., 4., 5., 1., 4., -1.]
      indx = [( k, k = 1, size(array) )]

      k = 2
      call arg_select(array, indx, k, kth_smallest)
      print*, array(kth_smallest) ! print 1.0

      k = 7
      ! Due to the previous call to arg_select, we know for sure this is in an
      ! index >= 2
      call arg_select(array, indx, k, kth_smallest, left=2)
      print*, array(kth_smallest) ! print 5.0

      k = 6
      ! Due to the previous two calls to arg_select, we know for sure this is in
      ! an index >= 2 and <= 7
      call arg_select(array, indx, k, kth_smallest, left=2, right=7)
      print*, array(kth_smallest) ! print 4.0

    end program demo_arg_select
```

## Comparison with using `sort`

The following program compares the timings of `select` and `arg_select` for

> computing the median of an array, vs using `sort` from stdlib.  In theory we

Reviewer: ```suggestion
computing the median of an array, vs using `sort` from `stdlib`.  In theory we
```

                call system_clock(t0)
                call select(local_random_vals, k, result_select)
                call system_clock(t1)
                time_select = time_select + (t1 - t0)

                ! Compute the median with arg_select, assuming N is odd
                local_random_vals = random_vals
                indx_local = indx
                call system_clock(t0)
                call arg_select(local_random_vals, indx_local, k, result_arg_select)
                call system_clock(t1)
                time_arg_select = time_arg_select + (t1 - t0)

                if(result_select /= result_sort) select_test_passed = .FALSE.
                if(local_random_vals(result_arg_select) /= result_sort) arg_select_test_passed = .FALSE.
            end do

            print*, "select    ; N=", N, '; ', merge('PASS', 'FAIL', select_test_passed), &
                '; Relative-speedup-vs-sort:', (1.0*time_sort)/(1.0*time_select)
            print*, "arg_select; N=", N, '; ', merge('PASS', 'FAIL', arg_select_test_passed), &
                '; Relative-speedup-vs-sort:', (1.0*time_sort)/(1.0*time_arg_select)

        end subroutine

  end program
```

> The results seem consistent with expectations when the `array` is large; the program prints:

Reviewer: Nice results ;)

---
title: Selection Procedures
---

# The `stdlib_selection` module

[TOC]

## Overview of selection

Suppose you wish to find the value of the kth-smallest entry in an array, or
the index of that value. While it could be done by sorting the whole array

> using `sort` or `sort_index` from `stdlib_sorting` and then finding the k-th

Reviewer: ```suggestion
using `[[stdlib_sorting(module):sort(interface)]]` or `[[stdlib_sorting(module):sort_index(interface)]]` from `[[stdlib_sorting(module)]]` and then finding the k-th
```

entry, that would require O(N x LOG(N)) time. However selection of a single

> entry can be done in O(N) time, so is much faster for large arrays.  This is

Reviewer: ```suggestion
entry can be done in O(N) time, which is much faster for large arrays.  This is
```

useful, for example, to quickly find the median of an array, or some other
percentile.

The Fortran Standard Library therefore provides a module, `stdlib_selection`,
which implements selection algorithms.

## Overview of the module

The module `stdlib_selection` defines two generic subroutines:
* `select` is used to find the kth-smallest entry of an array. The input
array is also modified in-place, and on return will be partially sorted
such that `all(array(1:k) <= array(k)))`  and `all(array(k) <= array((k+1):size(array)))`.
The user can optionally specify `left` and `right` indices to constrain the search
for the kth-smallest value. This can be useful if you have previously called `select`
to find a smaller or larger rank (that will have led to partial sorting of

> `array`, thus implying some constraint on the location).

Reviewer: ```suggestion
`array`, thus implying some constraints on the location).
```

* `arg_select` is used to find the index of the kth-smallest entry of an array.
In this case the input array is not modified, but the user must provide an
input index array with the same size as `array`, having unique indices from
`1:size(array)`, which is modified instead. On return the index array is modified
such that `all(array(index(1:k)) <= array(index(k)))` and `all(array(k) <= array(k+1:size(array)))`.
The user can optionally specify `left` and `right` indices to constrain the search
for the kth-smallest value. This can be useful if you have previously called `arg_select`
to find a smaller or larger rank (that will have led to partial sorting of

> `index`, thus implying some constraint on the location).

Reviewer: ```suggestion
`index`, thus implying some constraints on the location).
```

#### Licensing

The Fortran Standard Library is distributed under the MIT
License. However components of the library may be based on code with
additional licensing restrictions. In particular `select` and `arg_select`
were derived by modifying a matlab implementation of "qselect" by Manolis
Lourakis, https://www.mathworks.com/matlabcentral/fileexchange/68947-qselect

> Below is the licence of the matlab qselect

Reviewer: ```suggestion
Below is the license of the matlab qselect
```

> Copyright (c) 2018, Manolis Lourakis

Reviewer: Question: I am not an expert in licensing. Is this license compatible with the MIT license?

Author: I'm also not an expert on this. 

I'll try to email the matlab code author, and seek explicit permission to release the modified version under an MIT licence while still crediting the original implementation in the comments. A similar thing was done for the some of the sorting routines. 

If that's not possible, it won't be too difficult to develop an alternative implementation. The interface and initial checks in the pull-request are all original. It is just the "partition while-loop" that mirrors the matlab code. This is a pretty standard operation that could be replaced with a new implementation (e.g. informed by coretran, or something else).

Reviewer 4: Same, i'm not a lawyer.... Getting the permission from the author would allow a re-release under a different license than the original.  
The alternative might be to simply state that code written explicitly for stdlib is released under the MIT license, but some parts (translations like this) may have their own license.  I like this [image](https://dwheeler.com/essays/floss-license-slide.html) when referring to licenses.  This selection license looks more like a BSD3 license, and would therefore not be compatible with stdlib **If stdlib was released entirely under MIT**

I know there are stipulations with the less permissive GPL licenses that the entire library would have to be released under GPL, but I don't see any such stipulation in the MIT or BSD (Or this selection algo) text.  Does this not mean a library can be piece-by-piece licensed?  Maybe this should move to another thread?

Author: I just emailed the original author, to seek permission to release under MIT while citing the original work.

Reviewer: > I just emailed the original author, to seek permission to release under MIT while citing the original work. That will likely be easier than figuring out the legal details.

Good. Thank you.
Otherwise, I guess that some parts of stdlib `sort` could be used.

Author: I just implemented changes to the code so it draws on the Coretran implementation of @leonfoks rather than the matlab implementation. 

Coretran has a BSD3 licence -- actually this is very similar to the licence of the matlab code, so likely there would not have been problems. But it is good to make sure.

  names of its contributors may be used to endorse or promote products
  derived from this software without specific prior written permission.
THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE
FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

### Specifications of the `stdlib_selection` procedures

#### `select` - find the kth smallest value in an input array

##### Status

Experimental

##### Description

Returns the k-th smallest value of array `a(:)`, and also partially sorts `a(:)`

> such that `all(a(1:k) <= a(k))`  and `all(a(k) <= a((k+1):size(a)))`

Reviewer: I suggest to rename the array `a` to `array`.

Author: I've implemented this in a new commit.

##### Syntax

`call [[stdlib_selection(module):select(interface)]]( a, k, kth_smallest [, left, right ] )`

##### Class

Generic subroutine.

##### Arguments

`a` : shall be a rank one array of any of the types:
`integer(int8)`, `integer(int16)`, `integer(int32)`, `integer(int64)`,
`real(sp)`, `real(dp)`, `real(qp)`. It is an `intent(inout)` argument. On input it is
the array in which we search for the kth smallest entry. On return its elements
will be partially sorted such that:

>  `all( a(1:k-1) <= a(k) )` and `all(a(k) <= a(k+1:size(a)))`.

Reviewer: I suggest to move the sentence `On input ... a)))`. to the next section.

Author: I'm not sure about this - do you mean move it to the "Notes" section?

One thing I am conscious of is that some new users will be confused by the fact that `array` is changed, and I really want to make that point "hard to miss" in the documentation.

We could potentially just say `... On outputs its elements are reordered, see Notes below`. But I'm not sure if that is better than the current approach. Thoughts?

Reviewer: You mentioned it is an `intent(inout)`. So it is clear enough for me.

I think that the sentence I pointed out should be moved to the `result/output` section.

Author: I implemented this in some updates just now.

order), the code does not check for this.

Selection of a single value should have runtime of O(`size(a)`), so it is asymptotically faster
than sorting `a` entirely.

##### Example

```fortran
    program demo_arg_select
      use stdlib_selection, only: arg_select
      implicit none

      real, allocatable :: array(:)
      integer, allocatable :: indx(:)
      integer :: kth_smallest
      integer :: k, left, right

      array = [3., 2., 7., 4., 5., 1., 4., -1.]
      indx = (/( k, k = 1, size(array) )/)

      k = 2
      call arg_select(array, indx, k, kth_smallest)
      print*, array(kth_smallest) ! print 1.0

      k = 7
      ! Due to the previous call to arg_select, we know for sure this is in an
      ! index >= 2
      call arg_select(array, indx, k, kth_smallest, left=2)
      print*, array(kth_smallest) ! print 5.0

      k = 6
      ! Due to the previous two calls to arg_select, we know for sure this is in
      ! an index >= 2 and <= 7
      call arg_select(array, indx, k, kth_smallest, left=2, right=7)
      print*, array(kth_smallest) ! print 4.0

    end program demo_arg_select
```

> ## Comparison with using sort

Reviewer: ```suggestion
## Comparison with using `sort`
```
